test_runner/performance/test_pageserver.py

```diff
@@ -0,0 +1,177 @@
+import json
+import os
+import shutil
+import subprocess
+import time
+from pathlib import Path
+from typing import List, Tuple
+
+import pytest
+from fixtures.benchmark_fixture import NeonBenchmarker
+from fixtures.log_helper import log
+from fixtures.neon_fixtures import NeonEnv, NeonEnvBuilder, PgBin, last_flush_lsn_upload
+from fixtures.pageserver.utils import wait_until_tenant_active
+from fixtures.remote_storage import LocalFsStorage, RemoteStorageKind
+from fixtures.types import TenantId, TimelineId
+
+
+@pytest.fixture(scope="function")
+def snapshotting_env(
+    neon_env_builder: NeonEnvBuilder, pg_bin: PgBin, test_output_dir: Path
+) -> Tuple[NeonEnv, TimelineId, List[TenantId]]:
+    """
+    The fixture prepares environment or restores it from a snapshot.
+
+    The logic is the following:
+    - if the snapshot directory exists, the snapshot is restored from it
+    - if there is no snapshot, the environment is initialized from scratch and stored in a snapshot
+    - if the fixture is executed on CI (it has CI=true in the environment), the snapshot is not saved
+    """
+
+    snapshot_dir = test_output_dir.parent / f"snapshot-{test_output_dir.name}"
+    save_snapshot = os.getenv("CI", "false") != "true"
+
+    neon_env_builder.enable_generations = True
+    neon_env_builder.enable_pageserver_remote_storage(RemoteStorageKind.LOCAL_FS)
+
+    # create our template tenant
+    tenant_config_mgmt_api = {
+        "gc_period": "0s",
+        "checkpoint_timeout": "10 years",
+        "compaction_period": "20 s",
+        "compaction_threshold": 10,
+        "compaction_target_size": 134217728,
+        "checkpoint_distance": 268435456,
+        "image_creation_threshold": 3,
+    }
+
+    if snapshot_dir.exists():
+        env = neon_env_builder.from_repo_dir(snapshot_dir)
+        ps_http = env.pageserver.http_client()
+        tenants = list({TenantId(t.name) for t in (snapshot_dir.glob("pageserver_*/tenants/*"))})
+        template_timeline = env.initial_timeline
+
+        env.broker.try_start()
+
+        assert env.attachment_service is not None
+        env.attachment_service.start()
+
+        # Wait for the attachment service to start
+        time.sleep(5)
+
+        for tenant in tenants:
+            env.attachment_service.attach_hook_issue(tenant, 1)
```

Comment:
Hm, shouldn't the rehydration of the attachment service happen as part of `from_repo_dir`?

Reply:
Swapped most of the code with `neon_env_builder.start()`